src/main/java/com/powsybl/openloadflow/ac/ReactiveLimitsOuterLoop.java

```diff
@@ -115,30 +123,95 @@ private OuterLoopStatus switchPvPq(List<PvToPqBus> pvToPqBuses, EquationSystem e
         }
 
         if (!pvToPqBuses.isEmpty()) {
-            status = OuterLoopStatus.UNSTABLE;
+            done = true;
 
             for (PvToPqBus pvToPqBus : pvToPqBuses) {
                 // switch PV -> PQ
                 switchPvPq(pvToPqBus.bus, equationSystem, variableSet, pvToPqBus.qLimit);
-                LOGGER.trace("Switch bus {} PV -> PQ, q={} < {}Q={}", pvToPqBus.bus.getId(), pvToPqBus.q * PerUnit.SB,
+                LOGGER.trace("Switch bus '{}' PV -> PQ, q={} < {}Q={}", pvToPqBus.bus.getId(), pvToPqBus.q * PerUnit.SB,
                         pvToPqBus.limitDirection == ReactiveLimitDirection.MAX ? "max" : "min", pvToPqBus.qLimit * PerUnit.SB);
             }
         }
 
         LOGGER.info("{} buses switched PV -> PQ ({} bus remains PV}", pvToPqBuses.size(), modifiedRemainingPvBusCount);
 
-        return status;
+        return done;
+    }
+
+    private void switchPqPv(LfBus bus, EquationSystem equationSystem, VariableSet variableSet) {
+        bus.setVoltageControl(true);
+
+        Equation qEq = equationSystem.createEquation(bus.getNum(), EquationType.BUS_Q);
+        qEq.setActive(false);
+
+        LfBus controlledBus = bus.getControlledBus().orElse(null);
+        if (controlledBus != null) {
+            updateControlledBus(controlledBus, equationSystem, variableSet);
+        } else {
+            Equation vEq = equationSystem.createEquation(bus.getNum(), EquationType.BUS_V);
+            vEq.setActive(true);
+        }
+    }
+
+    private static final class PqToPvBus {
+
+        private final LfBus bus;
+
+        private final ReactiveLimitDirection limitDirection;
+
+        private PqToPvBus(LfBus bus, ReactiveLimitDirection limitDirection) {
+            this.bus = bus;
+            this.limitDirection = limitDirection;
+        }
+    }
+
+    private boolean switchPqPv(List<PqToPvBus> pqToPvBuses, EquationSystem equationSystem, VariableSet variableSet) {
+        int pqPvSwitchCount = 0;
```

Comment:
Can we just add a small description here: "If Q is equal to Qmin and V is less than targetV, it means that the PQ bus can be unlocked in order to increase the reactive power and reach its targetV. If Q is equal to Qmax and V is more than targetV, it means that the PQ bus can be unlocked in order to decerase the reactive power and reach its target V." With your own words of course.

Reply:
Added with your very clear wording!